contracts/l2/L2DAIWormholeBridge.sol

```diff
@@ -0,0 +1,124 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity >=0.7.6;
+pragma abicoder v2;
+
+import {iOVM_L1ERC20Bridge} from "@eth-optimism/contracts/iOVM/bridge/tokens/iOVM_L1ERC20Bridge.sol";
+import {iOVM_L2ERC20Bridge} from "@eth-optimism/contracts/iOVM/bridge/tokens/iOVM_L2ERC20Bridge.sol";
+import {OVM_CrossDomainEnabled} from "@eth-optimism/contracts/libraries/bridge/OVM_CrossDomainEnabled.sol";
+import {OVM_L2CrossDomainMessenger} from "@eth-optimism/contracts/OVM/bridge/messaging/OVM_L2CrossDomainMessenger.sol";
+import {WormholeGUID, WormholeLib} from "../common/LibWormholeGUID.sol";
+import {L1DAIWormholeBridge} from "../l1/L1DAIWormholeBridge.sol";
+
+interface Mintable {
+  function mint(address usr, uint256 wad) external;
+
+  function burn(address usr, uint256 wad) external;
+}
+
+contract L2DAIWormholeBridge is OVM_CrossDomainEnabled {
+  using WormholeLib for WormholeGUID;
+
+  // --- Auth ---
+  mapping(address => uint256) public wards;
+
+  function rely(address usr) external auth {
+    wards[usr] = 1;
+    emit Rely(usr);
+  }
+
+  function deny(address usr) external auth {
+    wards[usr] = 0;
+    emit Deny(usr);
+  }
+
+  modifier auth() {
+    require(wards[msg.sender] == 1, "L2DAITokenBridge/not-authorized");
+    _;
+  }
+
+  event Rely(address indexed usr);
+  event Deny(address indexed usr);
+
+  address public immutable l1Token;
+  address public immutable l2Token;
+  address public immutable l1DAITokenBridge;
+  bytes32 public immutable sourceDomain;
+  mapping(bytes32 => uint256) public batchedDaiToFlush;
+
+  event WormholeInitialized(WormholeGUID wormhole);
+  event Flushed(bytes32 targetDomain, uint256 dai);
+
+  constructor(
+    address _l2CrossDomainMessenger,
+    address _l2Token,
+    address _l1Token,
+    address _l1DAITokenBridge,
+    bytes32 _sourceDomain
+  ) public OVM_CrossDomainEnabled(_l2CrossDomainMessenger) {
+    wards[msg.sender] = 1;
+    emit Rely(msg.sender);
+
+    l2Token = _l2Token;
+    l1Token = _l1Token;
+    l1DAITokenBridge = _l1DAITokenBridge;
+    sourceDomain = _sourceDomain;
+  }
+
+  function initiateWormhole(
+    bytes32 targetDomain,
+    address receiver,
+    uint128 amount,
+    address operator
+  ) external {
+    WormholeGUID memory wormhole = WormholeGUID({
+      sourceDomain: sourceDomain,
+      targetDomain: targetDomain,
+      receiver: receiver,
+      operator: operator,
+      amount: amount,
+      nonce: uint64(OVM_L2CrossDomainMessenger(address(getCrossDomainMessenger())).messageNonce()), // gas optimization, we don't need to maintain our own nonce
+      timestamp: uint64(block.timestamp)
+    });
+
+    batchedDaiToFlush[targetDomain] += amount;
```

Comment:
I don't think there is an issue, but care should be taken if the `targetDomain` is not valid. There are two scenarios here:

1. It's a random string and the funds are lost.
2. It's set to something which will be launching later. Such as guessing we will add "STARKNET" in the future or something.

I don't think there is a problem in either case, but need to think on this.

Reply:
I think it's worth the cost of the check to prevent the loss of user funds. It would also prevent minting bad debt in case the oracles were to sign wormholes for invalid domains (e.g. we offboard a domain but the oracle API fails to sync their list of valid domains).